### PR TITLE
🐙 Unlock Android Studio integration

### PR DESCRIPTION
**Pull request by simoarpe**
## Resolves https://github.com/brave/brave-browser/issues/31782

kraken

**It's time to release the Kraken and fix Android Studio integration for Brave codebase.**

## How To
Pull this branch and follow the steps below to test the Android Studio integration.

* Run `npm run sync`.
* Build Brave with `npm run build ...`
* Make sure the working directory is  `brave-browser/src/brave` - This step is required, or the script won't work.
* Run `npm run gen_gradle -- Debug --target_arch=x86`
* The script will generate a `gradle` folder inside the ninja generation folder (e.g. `out/android_Debug_x86/gradle`).
* When the script has completed, open Android Studio and select "Import Project"
* Select the folder containing the generated project e.g. `out/android_Debug_x86/gradle`.
* Enjoy!

## Regenerate Gradle Build Files

You'll need to re-run `npm run build` and `npm run gen_gradle -- Debug --target_arch=x86` whenever new directories containing source files are added.

After regenerating, Android Studio should prompt you to “Sync”. If it doesn't, try some of the following options:
* File -> “Sync Project with Gradle Files”
* Button with two arrows on the right side of the top strip.
* Help -> Find Action -> “Sync Project with Gradle Files”
* `npm run sync` && `npm run build ...`
* File -> “Invalidate Caches / Restart...”

To delve more into the topic -> https://chromium.googlesource.com/chromium/src.git/+/main/docs/android_studio.md

## Android Studio

When opening the project for the first time, select `Project Files` to have a better overview of all available projects.

image (6)

The most important relevant projects for daily work are two:

* `brave/brave-browser/src/out/android_Debug_x86/gen/brave` - Mojo generated and native files 
* `brave/brave-browser/src/brave` - Java source code of Brave app

## Bugs, Issues & Improvements

🙏 Please report any bug, issue or suggestion to @simoarpe 

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by goodov** [comment]
1. please migrate all patches to `import_inline` and `override_utils` approach. Your goal is to have a single-line patch and have all brave-specific changes be in a separate file in `brave/chromium_src` dir. Example:
   https://github.com/brave/brave-core/blob/v1.58.46/chromium_src/tools/symsrc/source_index.py 
  https://github.com/brave/brave-core/blob/v1.58.46/patches/tools-symsrc-source_index.py.patch
2. add `npm run gen_gradle` command to make gradle project generation as simple as possible.

when you will call it via `npm run` command, `PYTHONPATH` will already be set, so you don't need to add `sys.path` overrides.

File: patches/chrome-android-java-AndroidManifest.xml.patch [hunk 1]
```diff
@@ -58,23 +58,15 @@ index 6f3722cd64029848c1737ced63889cf7cb41c5aa..266f72ac373c9db6834f4586a88d6234
  
      {% block extra_uses_permissions %}
      {% endblock %}
-@@ -181,6 +194,7 @@ by a child template that "extends" this file.
-     <!-- Set android:largeHeap to "true" to allow more than the default
-          Java heap limit (32Mb on Nexus S, 48Mb on Xoom). -->
-       <application android:name="{% block application_name %}org.chromium.chrome.browser.base.SplitChromeApplication{% endblock %}"
-+        android:requestLegacyExternalStorage="true"
```
Comment (bridiver) [inline patches/chrome-android-java-AndroidManifest.xml.patch:65]: why is this removed?

File: patches/build-util-android_chrome_version.py.patch [hunk 2]
```diff
@@ -1,16 +1,36 @@
...
+ 
++import os.path
++import sys
++
++_BUILD_ANDROID = os.path.join(os.path.dirname(__file__), os.pardir, 'android')
++sys.path.append(_BUILD_ANDROID)
++
++from pylib import constants
++from pylib.constants import host_paths
++_BRAVE_SCRIPT_PATH = os.path.join(host_paths.DIR_SOURCE_ROOT, 'brave',
```
Comment (bridiver) [inline patches/build-util-android_chrome_version.py.patch:17]: this shouldn't be necessary, this should already be added to python path

File: patches/build-util-android_chrome_version.py.patch [hunk 3]
```diff
@@ -1,16 +1,36 @@
...
 +++ b/build/util/android_chrome_version.py
-@@ -199,6 +199,7 @@ def _GetAbisToDigitMask(build_number):
+@@ -48,6 +48,19 @@ For example:
+ import argparse
+ from collections import namedtuple
+ 
++import os.path
++import sys
++
++_BUILD_ANDROID = os.path.join(os.path.dirname(__file__), os.pardir, 'android')
```
Comment (bridiver) [inline patches/build-util-android_chrome_version.py.patch:12]: same as below, these are added in config.js

File: patches/build-util-android_chrome_version.py.patch [hunk 4]
```diff
@@ -1,16 +1,36 @@
...
-index f00cd99d2bfa5793d6ed5d96025885ad83af412f..4e49484e6bdb3648739521c6125fbacb8516c33c 100755
+index f00cd99d2bfa5793d6ed5d96025885ad83af412f..41f44538d04232f990d0af59682b92f43fe1fa23 100755
 --- a/build/util/android_chrome_version.py
 +++ b/build/util/android_chrome_version.py
-@@ -199,6 +199,7 @@ def _GetAbisToDigitMask(build_number):
+@@ -48,6 +48,19 @@ For example:
+ import argparse
+ from collections import namedtuple
+ 
++import os.path
```
Comment (bridiver) [inline patches/build-util-android_chrome_version.py.patch:9]: I'd like to find some other way to do this like maybe an android studio config file? Or @goodov maybe we should change our python path and other overrides into a file you can source from the shell?

Comment (bridiver) [reply patches/build-util-android_chrome_version.py.patch:9]: maybe have it generate an env/env.bat file at runtime

**Comment by simoarpe** [comment]
Thank you for the suggestion @goodov, I'll ping you once ready for another review round!

File: patches/chrome-android-java-AndroidManifest.xml.patch (hunk 1 above)
Comment (simoarpe) [reply patches/chrome-android-java-AndroidManifest.xml.patch:65]: I've removed it as it was adding this attribute multiple times in many `AndroidStudio.xml` files.
Consequently, the Gradle build generation was failing as each Android Studio XML file was parsed to generate the right build options.
I'm attaching a screenshot of the error message of the original Python script. The error message was not very explanatory, so I ended up adding extra logs and after some digging it turned out that `android:requestLegacyExternalStorage="true"` was the duplicated attribute.

I'm going to check better and see if we can take a safer approach here. (Don't want to risk of leaving out the attribute for some files...)

image__7_

**Comment by simoarpe** [comment]
Feedback addressed! 
Ready for another review round @goodov , @bridiver.

File: chromium_src/build/android/gradle/generate_gradle.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+import override_utils
+
+_GENERATED_JAVA_SUBDIR = 'generated_java'
+_INPUT_SRC_JARS = 'input_srcjars'
+_BRAVE_GEN_COMPONENTS = os.path.join('gen', 'brave', 'components')
+
+
+@override_utils.override_method(_ProjectEntry)
+def GeneratedJavaSubdir(orig_func, self):
```
Comment (goodov) [inline chromium_src/build/android/gradle/generate_gradle.py:16]: rename `orig_func` -> `_` if you're not using it.

File: chromium_src/build/android/gradle/generate_gradle.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+import os
+
+import override_utils
+
+_GENERATED_JAVA_SUBDIR = 'generated_java'
+_INPUT_SRC_JARS = 'input_srcjars'
+_BRAVE_GEN_COMPONENTS = os.path.join('gen', 'brave', 'components')
+
+
```
Comment (goodov) [inline chromium_src/build/android/gradle/generate_gradle.py:14]: add `# pylint: disable=undefined-variable`

File: patches/build-android-gradle-generate_gradle.py.patch [hunk 7]
```diff
@@ -0,0 +1,12 @@
+diff --git a/build/android/gradle/generate_gradle.py b/build/android/gradle/generate_gradle.py
+index bc05baf9bbf34e5c0a41eef6198932d4f1ba2afd..2d293fb7f4ecbf2be78562fa05109c25b18a4b81 100755
+--- a/build/android/gradle/generate_gradle.py
++++ b/build/android/gradle/generate_gradle.py
+@@ -932,6 +932,6 @@ def main():
+   print('For more tips: https://chromium.googlesource.com/chromium/src.git/'
+         '+/main/docs/android_studio.md')
+ 
+-
```
Comment (goodov) [inline patches/build-android-gradle-generate_gradle.py.patch:9]: don't remove this empty line from the original file.

File: build/commands/lib/genGradle.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+const config = require('../lib/config')
+const util = require('../lib/util')
+const Log = require('./logging')
+
+let options = config.defaultOptions
+options.continueOnFail = false
+
+const genGradle = (passthroughArgs, build_dir) => {
+        Log.progressStart('Generating Gradle files')
```
Comment (goodov) [inline build/commands/lib/genGradle.js:14]: pls use this instead: `Log.progessScope('Generating Gradle files', () => { ... })`

also pls apply formatting using Prettier (you can install it into vscode and run manually).

File: build/commands/lib/genGradle.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+const config = require('../lib/config')
+const util = require('../lib/util')
```
Comment (goodov) [inline build/commands/lib/genGradle.js:7]: `./config`
`./util`

File: build/commands/lib/genGradle.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+let options = config.defaultOptions
+options.continueOnFail = false
+
+const genGradle = (passthroughArgs, build_dir) => {
+  Log.progressScope('Generating Gradle files', () => {
+    args = [
+      'build/android/gradle/generate_gradle.py',
+      '--output-directory',
+      build_dir
```
Comment (goodov) [inline build/commands/lib/genGradle.js:18]: the output directory can be derived the same way we generate `android_Component_arm64` and other similar dirs.

so it could be nice if a command looks like `build, test` commands:
`npm run gen_gradle Component -- --target=arm64`

wdyt?

Comment (simoarpe) [reply build/commands/lib/genGradle.js:18]: I thought about that, wouldn't be a bit confusing considering that is something valid only for Android (so `target_os=android`) ? Also, `--output-directory` would be the main command, but there are a few other useful options like -v for verbose mode, or a specific sdk path with --sdk-path. These could be handy for debugging purposes.

Comment (simoarpe) [reply build/commands/lib/genGradle.js:18]: image

Comment (goodov) [reply build/commands/lib/genGradle.js:18]: > I thought about that, wouldn't be a bit confusing considering that is something valid only for Android (so `target_os=android`) ? Also, `--output-directory` would be the main command, but there are a few other useful options like -v for verbose mode, or a specific sdk path with --sdk-path. These could be handy for debugging purposes.

sure, but we can pass other parameters as is. We need to only "generate" `--output-directory` one using our known approach. Thankfully `--target_arch` is not used by `generate_gradle.py`.

Comment (simoarpe) [reply build/commands/lib/genGradle.js]: I've implemented a possible solution in the commit [`ad22da1` (#19385)](https://github.com/brave/brave-core/pull/19385/commits/ad22da1f25d5f8acb6d33be4009eaa834a69d05f).

This approach feels a bit convoluted, but maybe it's just me that was used to type `--output-directory` easily. This approach also required some trickery because the `allowUnknownOption` was also including the `--target_arch` parameter and the Gradle script was failing, so I ended up filtering the  `passthroughArgs`.

Let me know what you think!

Comment (goodov) [reply build/commands/lib/genGradle.js]: > I've implemented a possible solution in the commit [`ad22da1` (#19385)](https://github.com/brave/brave-core/pull/19385/commits/ad22da1f25d5f8acb6d33be4009eaa834a69d05f).
> 
> This approach feels a bit convoluted, but maybe it's just me that was used to type `--output-directory` easily. This approach also required some trickery because the `allowUnknownOption` was also including the `--target_arch` parameter and the Gradle script was failing, so I ended up filtering the `passthroughArgs`.
> 
> Let me know what you think!

logged https://github.com/brave/brave-browser/issues/31944

**Review by bsclifton — APPROVED** [review]
JS changes LGTM (adding npm tasks to package.json)

File: patches/chrome-android-java-AndroidManifest.xml.patch (hunk 1 above)
Comment (bridiver) [reply patches/chrome-android-java-AndroidManifest.xml.patch]: cc @SergeyZhukovsky

Comment (bridiver) [reply patches/chrome-android-java-AndroidManifest.xml.patch]: Also it seems odd to me that this would fail for us when it's part of upstream code. Why does this fail for us, but not for chromium?

Comment (SergeyZhukovsky) [reply patches/chrome-android-java-AndroidManifest.xml.patch]: yeah that's strange, maybe upstream team doesn't generate it or it's not applied for the tests, I think we are ok as far as it's not in the main manifest file.

**Review by bridiver — APPROVED** [review]
